include/envoy/buffer/buffer.h

```diff
@@ -215,8 +236,26 @@ class Instance {
     constexpr const auto displacement = Endianness == ByteOrder::BigEndian ? sizeof(T) - Size : 0;
 
     auto result = static_cast<T>(0);
-    copyOut(start, Size, reinterpret_cast<char*>(std::addressof(result)) + displacement);
-    return fromEndianness<Endianness>(result);
+    auto bytes = reinterpret_cast<char*>(std::addressof(result));
+    copyOut(start, Size, &bytes[displacement]);
+
+    constexpr const auto most_significant_read_byte =
+        Endianness == ByteOrder::BigEndian ? displacement : Size - 1;
+
+    constexpr const auto all_bits_enabled = static_cast<T>(~static_cast<T>(0));
+
+    const auto extension =
+        std::is_signed<T>::value && Size < sizeof(T) && bytes[most_significant_read_byte] < 0
+            ? Endianness == ByteOrder::BigEndian
+                  ? static_cast<T>(
+                        static_cast<typename std::make_unsigned<T>::type>(all_bits_enabled) >>
+                        (Size * CHAR_BIT))
+                  : static_cast<T>(
+                        static_cast<typename std::make_unsigned<T>::type>(all_bits_enabled)
+                        << (Size * CHAR_BIT))
+            : static_cast<T>(0);
+
+    return fromEndianness<Endianness>(static_cast<T>(result | extension));
```

Comment:
Doesn't this only work because the machines we're running tests on are little-endian?

On an LE host, peekBE<int32_t, 3> with a buffer containing F1 02 03 would produce result == 0x0302F100 (because BE to host conversion has not happened yet) and extension == 0xFF. This yields result | extension == 0x0302F1FF, which, after conversion is 0xFFF10203. That's correct.

On a BE host, peekBE<int32_t, 3> with a buffer containing F1 02 03 would produce result == 0x00F10203 (because the host is BE), but extension == 0xFF yields result | extension == 0x00F102FF. 

I think the correct implementation would be to always compute extension without regard to endianness (as the correct number of high order bits) and apply it *after* result has been converted to the host byte order:

```
  constexpr const auto extension = 
      std::is_signed<T>::value && Size < sizeof(T) && bytes[most_significant_read_byte] < 0
          ? static_cast<T>(
                  static_cast<typename std::make_unsigned<T>::type>(all_bits_enabled) << (Size * CHAR_BIT))
          : static_cast<T>(0);

  // this may require additional casting:
  return fromEndianness<Endianness>(static_cast<T>(result)) | extension;
```

Reply:
you're absolutely correct. while helping Jason out yesterday debugging some tests I've realized my brain fart on the initial sign extension code

Reply:
ah, this was my bad, @juchem actually pointed it out yesterday but I ended up adding it back when working on big endian test cases, without realizing the host endianness factor. I'll correct it now. Thanks!

Reply:
actually, there's an additional detail that @zuercher pointed out
we must merge the extension bits after the endianness conversion, not before (move ` | extension` outside of the call to `fromEndianness`, as specified in his code snippet)